kura/org.eclipse.kura.driver.opcua.provider/src/main/java/org/eclipse/kura/internal/driver/opcua/OpcUaDriver.java

```diff
@@ -451,61 +399,93 @@ public void unregisterDriverListener(final DriverListener listener) throws Conne
+                    logger.warn(message.writeFailed(), e);
+                }
                 record.setTimestamp(System.currentTimeMillis());
-                continue;
-            }
+            });
+        }
+        return Collections.unmodifiableList(records);
+    }
+
+    private static class OpcUaRequestInfo {
+
+        private final DataType dataType;
+        private final int nodeNamespaceIndex;
+        private final String nodeId;
+
+        public OpcUaRequestInfo(DataType dataType, int nodeNamespaceIndex, String nodeId) {
+            this.dataType = dataType;
+            this.nodeNamespaceIndex = nodeNamespaceIndex;
+            this.nodeId = nodeId;
+        }
+
+        private static void fail(DriverRecord record, String message) {
+            record.setDriverStatus(new DriverStatus(DRIVER_ERROR_CHANNEL_NOT_ACCESSIBLE, message, null));
+            record.setTimestamp(System.currentTimeMillis());
+        }
+
+        public static Optional<OpcUaRequestInfo> extract(DriverRecord record) {
+            final Map<String, Object> channelConfig = record.getChannelConfig();
             int nodeNamespaceIndex;
+            String nodeId;
+
             try {
-                nodeNamespaceIndex = Integer.parseInt(channelConfig.get(NODE_NAMESPACE_INDEX).toString());
-            } catch (final NumberFormatException nfe) {
-                record.setDriverStatus(new DriverStatus(DRIVER_ERROR_CHANNEL_NOT_ACCESSIBLE,
-                        message.errorRetrievingNodeNamespace(), null));
-                record.setTimestamp(System.currentTimeMillis());
-                continue;
+                nodeId = channelConfig.get(NODE_ID).toString();
+            } catch (final Exception e) {
+                fail(record, message.errorRetrievingNodeId());
+                return Optional.empty();
             }
-            final TypedValue<?> value = record.getValue();
-            final NodeId nodeId = new NodeId(nodeNamespaceIndex, channelConfig.get(NODE_ID).toString());
-            final VariableNode node = this.client.getAddressSpace().createVariableNode(nodeId);
-            final DataValue newValue = new DataValue(new Variant(value.getValue()));
+
             try {
-                node.writeValue(newValue).get();
-                record.setDriverStatus(new DriverStatus(WRITE_SUCCESSFUL));
+                nodeNamespaceIndex = Integer.parseInt(channelConfig.get(NODE_NAMESPACE_INDEX).toString());
             } catch (final Exception e) {
-                record.setDriverStatus(new DriverStatus(WRITE_FAILURE, message.writeFailed(), e));
+                fail(record, message.errorRetrievingNodeNamespace());
+                return Optional.empty();
             }
-            record.setTimestamp(System.currentTimeMillis());
+
+            final DataType dataType = (DataType) record.getChannelConfig().get(CHANNEL_VALUE_TYPE.value());
```

Comment:
replace `record.getChannelConfig()` with `channelConfig` local variable that is already present.